### PR TITLE
NCC: Lethal Scheme and ConniveEffect improvements

### LINKED ISSUE #373: Lethal Scheme

### PR DESCRIPTION
**Pull request by Northmoc**
closes #373

### CODE REVIEW

File: forge-game/src/main/java/forge/game/ability/effects/ConniveEffect.java [hunk 1]
```diff
@@ -50,21 +50,27 @@ public void resolve(SpellAbility sa) {
         moveParams.put(AbilityKey.LastStateBattlefield, sa.getLastStateBattlefield());
         moveParams.put(AbilityKey.LastStateGraveyard, sa.getLastStateGraveyard());
 
-        for (final Card c : getTargetCards(sa)) {
-            final Player p = c.getController();
+        CardCollection toConnive = getTargetCards(sa);
+        Card conniver;
+        while (toConnive.size() > 0) {
+            if (toConnive.size() > 1) {
+                conniver = hostCon.getController().chooseSingleEntityForEffect(toConnive, sa, "Choose conniver", null);
```
Comment (Northmoc) [inline forge-game/src/main/java/forge/game/ability/effects/ConniveEffect.java:57]: probably need to do a Localizer label here

**Comment by Hanmac** [comment]
> When multiple creatures are instructed to connive simultaneously, they connive one at a time in the order of their controller’s choice.

currently there isn't an effect that makes multiple creatures from different players connive

**Comment by Hanmac** [comment]
`discardedMap.put(p, toBeDiscarded);` probably needs some cleanup too

**Comment by tool4ever** [comment]
What happens when you cast this with convoke first, then flashback it but pay normally?

From the code, I'm a bit concerned it might still find the creatures on the second time and connive them again

**Comment by Hanmac** [comment]
> What happens when you cast this with convoke first, then flashback it but pay normally?
> 
> From the code, I'm a bit concerned it might still find the creatures on the second time and connive them again

`Creature.convoked` would be safe enough i think. imo better would be to ask the Spell instead and go over `Defined$ Convoked` or something like that

**Comment by Northmoc** [comment]
> What happens when you cast this with convoke first, then flashback it but pay normally?
> 
> From the code, I'm a bit concerned it might still find the creatures on the second time and connive them again

Tested this after making at least some of the other required changes, and nothing connives.

**Comment by Hanmac** [comment]
@Northmoc i found one: `Mask of the Schemer` and some others can make a creature you don't control convive
in that case the creatures controller does draw and discard.

with my comment i mean it more like there isn't a card yet that makes multiple player connive at the same time.

`Obscura Confluence` for example could make multiple creatures connive, but only one at a time

File: forge-game/src/main/java/forge/game/ability/effects/ConniveEffect.java (hunk 1 above)
Comment (Northmoc) [reply forge-game/src/main/java/forge/game/ability/effects/ConniveEffect.java]: See what you think of this!

File: forge-game/src/main/java/forge/game/ability/effects/ConniveEffect.java [hunk 2]
```diff
@@ -53,29 +52,29 @@ public void resolve(SpellAbility sa) {
...
-        CardCollection toConnive = getTargetCards(sa);
-        Card conniver;
-        while (toConnive.size() > 0) {
-            if (toConnive.size() > 1) {
-                conniver = hostCon.getController().chooseSingleEntityForEffect(toConnive, sa, "Choose conniver", null);
-            } else {
-                conniver = toConnive.get(0);
-            }
+        CardCollectionView toConnive = getTargetCards(sa);
+        final Player p = toConnive.getFirst().getController(); // currently all connivers will have the same controller
```
Comment (Hanmac) [inline forge-game/src/main/java/forge/game/ability/effects/ConniveEffect.java:56]: need fail check if toConnive is empty

**Comment by Hanmac** [comment]
Good thing it wasn't merged yet:
https://twitter.com/Dunkatog/status/1527461216603590680?s=20&t=bSC9kojHU1tRGFjqjsN7Lw

> You choose one and it connives. Then you choose the next one, and it connives, and so on. For more details, rule 701.47d should have you covered.

> 701.47d  If multiple permanents are instructed to connive at the same time, the first player in APNAP order who controls one or more of those permanents chooses one of them and it connives. Then if any permanents remain on the battlefield which have been instructed to connive and have not done so, this process is repeated.

so in theory: we should get the controllers, loop them and then loop the connive creatures ... sadly no nice order dialog :(

**Comment by Northmoc** [comment]
> Good thing it wasn't merged yet: https://twitter.com/Dunkatog/status/1527461216603590680?s=20&t=bSC9kojHU1tRGFjqjsN7Lw
> 
> > You choose one and it connives. Then you choose the next one, and it connives, and so on. For more details, rule 701.47d should have you covered.
> 
> > 701.47d  If multiple permanents are instructed to connive at the same time, the first player in APNAP order who controls one or more of those permanents chooses one of them and it connives. Then if any permanents remain on the battlefield which have been instructed to connive and have not done so, this process is repeated.
> 
> so in theory: we should get the controllers, loop them and then loop the connive creatures ... sadly no nice order dialog :(

I guess it saves users needing to memorize an order or wrangling the UI to highlight the current creature being connived... for now each Connive instance will have all connivers with just one controller, so we can skip the player loop (Obscura Confluence uses different Connive instance for each target)

**Comment by Hanmac** [comment]
ugh, i was thinking about Lethal Scheme, if you lose control of a creature that was used for convoke,
then it should convoke under a different player. for this we need to handle the the player in APNAP order

**Comment by Agetian** [comment]
Is this branch ready or is more work planned for it according to the considerations outlined above?

**Comment by Hanmac** [comment]
i asked again if the discard and counters happen at the same time or in sequence for our trigger/replacements

**Comment by Northmoc** [comment]
> i asked again if the discard and counters happen at the same time or in sequence for our trigger/replacements

I wondered about that too... his tweet made me think that maybe they don't happen all at once ☹️

File: forge-game/src/main/java/forge/game/ability/effects/ConniveEffect.java [hunk 3]
```diff
@@ -91,7 +91,6 @@ public void resolve(SpellAbility sa) {
                 }
 
                 CardCollection validCards = CardLists.getValidCards(dPHand, "Card", p, host, sa);
```
Comment (Hanmac) [inline forge-game/src/main/java/forge/game/ability/effects/ConniveEffect.java:93]: why was this filter there, after checking for non token, shouldn't all of them be Card ?

File: forge-game/src/main/java/forge/game/ability/effects/ConniveEffect.java [hunk 4]
```diff
@@ -65,20 +65,20 @@ public void resolve(SpellAbility sa) {
...
 
         for (final Player p : controllers) {
-            GameEntityCounterTable table = new GameEntityCounterTable();
-            final CardZoneTable triggerList = new CardZoneTable();
-            List<Card> allToDiscard = new ArrayList<>();
-            Map<Player, CardCollectionView> discardedMap = Maps.newHashMap();
-            Map<AbilityKey, Object> moveParams = AbilityKey.newMap();
-            moveParams.put(AbilityKey.LastStateBattlefield, sa.getLastStateBattlefield());
-            moveParams.put(AbilityKey.LastStateGraveyard, sa.getLastStateGraveyard());
             CardCollection connivers = new CardCollection();
```
Comment (Hanmac) [inline forge-game/src/main/java/forge/game/ability/effects/ConniveEffect.java:68]: CardLists.filterByController(p)

File: forge-game/src/main/java/forge/game/ability/effects/ConniveEffect.java (hunk 3 above)
Comment (Northmoc) [reply forge-game/src/main/java/forge/game/ability/effects/ConniveEffect.java]: Or maybe DiscardEffect needs a hard look, too 😆

File: forge-game/src/main/java/forge/game/ability/effects/ConniveEffect.java [hunk 5]
```diff
@@ -79,15 +79,15 @@ public void resolve(SpellAbility sa) {
 
                 p.drawCards(num, sa, moveParams);
 
-                CardCollectionView validCards = p.getCardsIn(ZoneType.Hand);
-                validCards = CardLists.filter(validCards, CardPredicates.Presets.NON_TOKEN);
-                if (validCards.isEmpty() || !p.canDiscardBy(sa, true)) { // hand being empty unlikely, just to be safe
+                CardCollection validDisards = (CardCollection) p.getCardsIn(ZoneType.Hand);
+                validDisards = CardLists.filter(validDisards, CardPredicates.Presets.NON_TOKEN);
```
Comment (Hanmac) [inline forge-game/src/main/java/forge/game/ability/effects/ConniveEffect.java:83]: so you don't need to cast:

`CardLists.filter(p.getCardsIn(ZoneType.Hand), CardPredicates.Presets.NON_TOKEN)`

**Review by Hanmac — APPROVED** [review]
👍

**Pull request merged**